src/Simplex/Chat/Store/Messages.hs

```diff
@@ -1067,6 +1068,46 @@ getDirectChatBefore_ db user@User {userId} ct@Contact {contactId} beforeChatItem
           |]
           (userId, contactId, search, beforeChatItemId, count)
 
+getDirectChatAround_ :: DB.Connection -> User -> Contact -> ChatItemId -> Int -> String -> IO (Chat 'CTDirect)
+getDirectChatAround_ db user@User {userId} ct@Contact {contactId} aroundItemId count search = do
+  let stats = ChatStats {unreadCount = 0, minUnreadItemId = 0, unreadChat = False}
+  chatItemIds <- getDirectChatItemsIdsAround_
+  currentTs <- getCurrentTime
+  chatItems <- mapM (safeGetDirectItem db user ct currentTs) chatItemIds
+  pure $ Chat (DirectChat ct) (reverse chatItems) stats
+  where
+    getDirectChatItemsIdsAround_ :: IO [ChatItemId]
+    getDirectChatItemsIdsAround_ = do
+      let (fetchCountBefore, fetchCountAfter) = divideFetchCountAround_ count
+      map fromOnly
+        <$> DB.queryNamed
+          db
+          [sql|
+            (SELECT chat_item_id
+            FROM chat_items
+            WHERE user_id = :userId AND contact_id = :contactId AND item_text LIKE '%' || :search || '%'
+              AND chat_item_id < :aroundItemId
+            ORDER BY created_at DESC, chat_item_id DESC
+            LIMIT :fetchCountBefore)
+            UNION ALL
+            (SELECT chat_item_id
+            FROM chat_items
+            WHERE user_id = :userId AND contact_id = :contactId AND item_text LIKE '%' || :search || '%'
+              AND chat_item_id >= :aroundItemId
+            ORDER BY created_at ASC, chat_item_id ASC
+            LIMIT :fetchCountAfter)
+            ORDER BY created_at DESC, chat_item_id DESC
+            LIMIT :count
+          |]
+          [ ":userId" := userId,
+            ":contactId" := contactId,
+            ":search" := search,
+            ":aroundItemId" := aroundItemId,
+            ":fetchCountBefore" := fetchCountBefore,
+            ":fetchCountAfter" := fetchCountAfter,
+            ":count" := count
+          ]
+
```

Comment:
to have less queries overall:

beforeIds <- getDirectChatItemsIdsBefore_
afterIds <- getDirectChatItemIdsAfter_
let idsToGet = beforeIds <> aroundItemId <> afterIds
mapM getItem idsToGet